### DISCUSSION TITLE
allowedVersions for maven artefacts

### DISCUSSION THREAD

**Discussion by swiss-chris** [renovatebot/renovate#20282] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

For the following config:
I would like to only accept dependency updates for dependencies who's versions fulfil the following 2 requirements:

1. Enforce SemVer with three digits (since Apache switched from yyyyMMdd based versioning without changing the artifact ID. Therefore older versions would qualify as newer ones).
2. Don't accept versions that contain ".redhat-.*" at the end of the version number.

I'm assuming Renovate doesn't understand Regex Lookaheads.

Does anyone know of a solution that works ?

### What you tried so far.

```
{
  groupName: "Apache Commons",
  packagePatterns: [
    "^org.apache.commons:",
    "^commons-beanutils:",
    "^commons-cli:",
    "^commons-collections:",
    "^commons-io:"
  ],
  "allowedVersions": "/^(\\d{1,3}\\.)(\\d*\\.)*\\d*(?!.*\\.redhat-.*)$/"
},
```

I'm assuming Renovate doesn't understand Regex Lookaheads.

### Relevant debug logs

<details><summary>Logs</summary>

```
INFO: Repository has invalid config (repository=...)
"error": {
"validationSource": "config",
"validationError": "Invalid allowedVersions",
"validationMessage": "The following allowedVersions does not parse as a valid version or range: "/^(\\d{1,3}\\.)(\\d*\\.)\\d(?!.\\.redhat-.)$/"",
"message": "config-validation",
"stack": "Error: config-validation\n at filterVersions (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/filter.ts:97:21)\n at lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:234:44)\n at processTicksAndRejections (internal/process/task_queues.js:95:5)\n at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:58:15)\n at /usr/src/app/node_modules/p-map/index.js:57:22"
}
WARN: Cannot ensure issue (repository=...)
"title": "Action Required: Fix Renovate Configuration"
INFO: Repository finished (repository=...)
"cloned": true,
```

</details>

**Comment by viceice** [comment]
yes, we're using re2 as regex engine. so no look ahead or behind.

**Comment by Churro** [comment]
1. To filter old versions of Apache commons, the `workarounds:mavenCommonsAncientVersion` preset could be helpful. See [here](https://docs.renovatebot.com/presets-workarounds/#workaroundsmavencommonsancientversion).
2. Excluding redhat strings may work using the "special negated regex syntax", see [here](https://docs.renovatebot.com/configuration-options/#allowedversions).

**Comment by swiss-chris** [comment]
@Churro thanks for your reply. I have searched multiple times, but I cannot find any documentation that explains how 2 rules are combined with Renovate. I was under the assumption that I had to pack all the logic into a single rule to get AND combination. 
In other words, if I have one rule `workarounds:mavenCommonsAncientVersion` and then another rule with `"allowedVersions": "!/.redhat-/"`, how are those 2 rules combined ? if the first one passes, are the dependencies updated  (OR logic) ? Or do all rules have to pass before a dependency is updated  (AND logic). I have looked multiple times for a good explanation of this, but so far have found only people asking questions about it and guessing... any help in this matter would be highly appreciated.

**Reply by Churro** [reply]
It depends on what you want to achieve. If you always strictly need both requirements, then the approach above may not suffice. In general, you can apply multiple package rules onto the same packages and by that, e.g. first filter ancient commons versions and then filter all versions not including the redhat string. Probably just a matter of trying out.

**Reply by swiss-chris** [reply]
Thanks. I'm still hoping one day I will find an official Renovate documentation that explains how it works.
